packages/playwright-test/src/loader.ts

```diff
@@ -324,6 +333,105 @@ ${'='.repeat(80)}\n`);
   }
 }
 
+class ProjectSuiteBuilder {
+  private _config: FullProjectInternal;
+  private _index: number;
+  private _testTypePools = new Map<TestTypeImpl, FixturePool>();
+  private _testPools = new Map<TestCase, FixturePool>();
+
+  constructor(project: FullProjectInternal, index: number) {
+    this._config = project;
+    this._index = index;
+  }
+
+  private _buildTestTypePool(testType: TestTypeImpl): FixturePool {
+    if (!this._testTypePools.has(testType)) {
+      const fixtures = this._applyConfigUseOptions(testType, this._config.use);
+      const pool = new FixturePool(fixtures);
+      this._testTypePools.set(testType, pool);
+    }
+    return this._testTypePools.get(testType)!;
+  }
+
+  // TODO: we can optimize this function by building the pool inline in cloneSuite
+  private _buildPool(test: TestCase): FixturePool {
+    if (!this._testPools.has(test)) {
```

Comment:
```suggestion
    if (this._testPools.has(test))
      return this._testPools.get(test)!;
    ...
```
nit: reduces nesting